File: frontend/src/lib/utils/smartTrim.ts
/**
 * Trims a string, removes all control characters, and collapses multiple
 * spaces into a single space, resulting in a clean and standardized string.
 *
 * @param input - The original string to be cleaned and trimmed.
 * @returns The resulting string after all cleaning operations are applied.
 */
export const smartTrim = (input: string): string => {
  // Trim and clear all control symbols
  let trimmedAndCleaned = input
      .trim()
      .replace(/[\p{C}]/gu, ''); 
     
  // Collapse multiple whitespace characters into a single space.
  let singleSpaced = trimmedAndCleaned.replace(/\s+/g, ' ');

  // Final trim in case the input started or ended with multiple spaces/control characters.
  return singleSpaced.trim();
}

File: backend/src/utils/smart_trim.rs
/// Trims a string, removes all control characters, and collapses multiple spaces into a single space, resulting in a clean and standardized string
pub fn smart_trim(input: &str) -> String {
    let trimmed_and_cleaned: String = input
        .trim()
        .chars()
        .filter(|c| !c.is_control())
        .collect();

    let mut result = String::with_capacity(trimmed_and_cleaned.len());
    let mut last_char_was_space = false;

    for c in trimmed_and_cleaned.chars() {
        if c.is_whitespace() {
            if !last_char_was_space {
                result.push(' ');
                last_char_was_space = true;
            }
        } else {
            result.push(c);
            last_char_was_space = false;
        }
    }

    result.trim().to_string()
}
